### PR TITLE
Migrate `std-core-cycle`, `obey-crate-type-flag`, `mixing-libs` and `issue-18943` `run-make` tests to `rmake.rs`

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

try-job: x86_64-apple-1
try-job: x86_64-msvc
try-job: aarch64-gnu

### CODE REVIEW

File: tests/run-make/mixing-libs/rmake.rs
Comment (fbstj) [inline tests/run-make/mixing-libs/rmake.rs:14]: I believe this should probably be 
```suggestion
    // Not putting `-C prefer-dynamic` here allows for static linking of librlib.rlib.
```
because I had to look back at the makefile to determine if it was prefer=dynamic (I didn't think it was)

Comment (Oneirical) [reply tests/run-make/mixing-libs/rmake.rs]: Good catch! Thanks.

File: src/tools/run-make-support/src/lib.rs
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:217]: Discussion: these APIs looks a bit weird to me for a couple of reasons:

- It's based on globbing which can be surprising if you don't document the edge cases, and even then it's not immediately obvious how the implementation behaves.
- It uses `dynamic_lib_name` et al. on constructed glob patterns, which seems a bit weird to me, I think it's cleaner if we extract out `dynamic_lib_extension` that's used by both `dynamic_lib_name` and the fs walker to determine what files to remove.
- The count APIs are *especially* weird because I think it's almost an anti-pattern, for vast majority of tests I would expected that we control and know *exactly* which output artifacts we generate. Not checking for those (e.g. by removing the expected output artifacts) almost smell like bugs.

Maybe take a look at how tidy does filewalking (e.g. <https://github.com/rust-lang/rust/blob/master/src/tools/tidy/src/walk.rs>) for inspiration, but we definitely don't want to directly mimic tidy's interfaces. 

I think it's easier to understand if we implement it in terms of `walkdir` or some other variant of explicit filesystem walk, and document if it is recursive or not.

What do you think?

File: src/tools/run-make-support/src/lib.rs
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:228]: Nit: appreciate not just `.unwrap()`-ing, maybe lowercase for consistency?

File: tests/run-make/lib-trait-for-trait-no-ice/rmake.rs
Comment (jieyouxu) [inline tests/run-make/lib-trait-for-trait-no-ice/rmake.rs:12]: Problem: I don't like relying on count here. For vast majority of tests, we should know *exactly* which output artifacts we generate and the expected exact output artifact not existing is a bug (in the test, in rustc, or both).

I would expect something more like

```rs
    fs_wrapper::remove_file(rust_lib_name("foo"));
```

File: tests/run-make/mixing-libs/rmake.rs
Comment (jieyouxu) [inline tests/run-make/mixing-libs/rmake.rs:19]: Discussion: I believe it's clearer and more precise with explicit `fs_wrapper::remove_file(dynamic_lib_name("rlib"))` and not glob removal.

File: tests/run-make/obey-crate-type-flag/rmake.rs
Comment (jieyouxu) [inline tests/run-make/obey-crate-type-flag/rmake.rs:13]: Question: do these pass if nothing are removed? If so, I think that's not very clear and I would prefer to either not have remove operations or to know exactly what is being removed here.

File: tests/run-make/obey-crate-type-flag/rmake.rs
Comment (jieyouxu) [inline tests/run-make/obey-crate-type-flag/rmake.rs:15]: Praise: I think this is a good use case of the count API.

**Review by jieyouxu — CHANGES_REQUESTED** [review]
Thanks for the PR. I have some concerns regarding the APIs and their implementation introduced in this PR that warrants further discussion, but the tests themselves look very reasonable.

**Comment by jieyouxu** [comment]
I've also took the liberty to edit the PR description and include try jobs for apple (darwin), windows and linux for the std-core-cycle. This PR should be run on those try jobs before approval.

@bors rollup=iffy (the std-core-cycle looks platform-difference sensitive)

**Comment by Oneirical** [comment]
@rustbot review
Needs try job.

**Comment by Oneirical** [comment]
I have removed glob once more and removed all instances of `remove_rlibs` and associated functions. I only kept a new version of `count_files_with_extension`.

File: src/tools/run-make-support/src/lib.rs
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:220]: Suggestion: I would recommend building this on top of `walkdir` also like in https://github.com/rust-lang/rust/pull/126427#discussion_r1643325225 so we don't swallow errors and if we do encounter errors, we report something that's actually meaningful for the tester.

Comment (jieyouxu) [reply src/tools/run-make-support/src/lib.rs:220]: @rustbot author

**Comment by Oneirical** [comment]
That's an entire wall of errors in tests completely unrelated to this one...

**Comment by jieyouxu** [comment]
Hopefully it's spurious (idk how that fails)

@bors try again

**Comment by jieyouxu** [comment]
Well that's concerning...

**Comment by jieyouxu** [comment]
Looks like a bug with how run-make-support's deps are loaded on windows, probably a bug within compiletest's `run_rmake_v2_test()` but I can only reliably debug this when I get access to Windows machine. Feel free to wait on this a bit if u want. @bors review

**Comment by Oneirical** [comment]
I think you meant @rustbot review

**Comment by Oneirical** [comment]
Would it be silly to bring over `find_files_by_prefix_and_extension` from #126957 while walkdir isn't fixed? I could do:

`find_files_by_prefix_and_extension(cwd(), "", "rlib").len()`

https://github.com/rust-lang/rust/pull/126957/files

**Comment by Kobzol** [comment]
Let's use the new helper function (the one that takes the closure) to avoid using `walkdir`, once https://github.com/rust-lang/rust/pull/126957 is merged.

**Comment by Oneirical** [comment]
Ah, those pesky submodules. This is going to take a little moment to fix. @rustbot author

**Comment by Kobzol** [comment]
I'll mark this as blocked on https://github.com/rust-lang/rust/pull/126957 because of the helper function.

@rustbot blocked

File: tests/run-make/lib-trait-for-trait-no-ice/rmake.rs
Comment (jieyouxu) [inline tests/run-make/lib-trait-for-trait-no-ice/rmake.rs:12]: Suggestion: this should be checking for existence of the lib file, not checking-by-proxy via trying to remove it I feel like?

File: tests/run-make/obey-crate-type-flag/rmake.rs
Comment (jieyouxu) [inline tests/run-make/obey-crate-type-flag/rmake.rs:15]: Suggestion: ditto here about checking for existence, not checking-by-proxy via trying to remove file.

**Review by jieyouxu — APPROVED** [review]
Looks good, r=Kobzol and I after addressing the suggestions.

**Comment by Oneirical** [comment]
Forgot to queue this one up.
@bors r=jieyouxu,kobzol

**Pull request merged**
